File: 10009 All Roads Lead Where?/solution.java
import java.io.*;
import java.util.*;

public class Main {
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StreamTokenizer st = new StreamTokenizer(br);
		st.nextToken();
		int cases = (int)st.nval;
		StringBuilder output = new StringBuilder();
		while(cases-- > 0) {
			st.nextToken();
			int m = (int)st.nval;
			st.nextToken();
			int n = (int)st.nval;
			int[] cityID = new int[26];
			Arrays.fill(cityID, -1);
			Edge[] edges = new Edge[m];
			ArrayList<String> cityName = new ArrayList<>();
			for(int i = 0; i < m; ++i) {
				st.nextToken();
				String city1 = st.sval;
				st.nextToken();
				String city2 = st.sval;
				if(cityID[city1.charAt(0) - 'A'] == -1) {
				    cityID[city1.charAt(0) - 'A'] = cityName.size();
				    cityName.add(city1);
				}
				if(cityID[city2.charAt(0) - 'A'] == -1) {
				    cityID[city2.charAt(0) - 'A'] = cityName.size();
				    cityName.add(city2);
				}
				edges[i] = new Edge(cityID[city1.charAt(0) - 'A'], cityID[city2.charAt(0) - 'A']);
			}
			ArrayList<Integer>[] conn = (ArrayList<Integer>[])new ArrayList[cityName.size()];
			for(int i = 0; i < conn.length; ++i)
				conn[i] = new ArrayList<>();
			for(Edge e : edges) {
				conn[e.parent].add(e.child);
				conn[e.child].add(e.parent);
			}
			while(n-- > 0) {
				st.nextToken();
				int start = cityID[st.sval.charAt(0) - 'A'];
				st.nextToken();
				int end = cityID[st.sval.charAt(0) - 'A'];
				int[] prev = new int[cityName.size()];
				boolean[] visited = new boolean[cityName.size()];
				visited[prev[start] = start] = true;
				ArrayList<Integer> curr = new ArrayList<>();
				curr.add(start);
				while(!visited[end]) {
					ArrayList<Integer> next = new ArrayList<>();
					for(int i = 0; i < curr.size(); ++i) {
						for(int j = 0; j < conn[curr.get(i)].size(); ++j) {
							if(!visited[conn[curr.get(i)].get(j)]) {
								next.add(conn[curr.get(i)].get(j));
								visited[conn[curr.get(i)].get(j)] = true;
								prev[conn[curr.get(i)].get(j)] = curr.get(i);
							}
						}
					}
					curr = new ArrayList<>(next);
				}
				ArrayList<Integer> route = new ArrayList<>();
				while(end != start) {
					route.add(end);
					end = prev[end];
				}
				output.append(cityName.get(start).charAt(0));
				for(int i = route.size() - 1; i >= 0; --i)
					output.append(cityName.get(route.get(i)).charAt(0));
				output.append('\n');
			}
			if(cases > 0)
				output.append('\n');
		}
		System.out.print(output);
	}
}

class Edge {
	int parent, child;
	Edge(int parent, int child) {
		this.parent = parent;
		this.child = child;
	}
}

File: 10009 All Roads Lead Where?/solution.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>

typedef struct {
    int parent, child;
} Edge;

int curr[25], next[25];
char cityName[26], city1[9], city2[9];

int main() {
    int cases;
    scanf("%d", &cases);
    while(cases--) {
        int m, n;
        scanf("%d %d ", &m, &n);
        int cityID[26];
        for(int i = 0; i < 26; ++i)
            cityID[i] = -1;
        Edge edges[m];
        int id = 0;
        for(int i = 0; i < m; ++i) {
            scanf("%s", city1);
            if(cityID[city1[0] - 'A'] == -1) {
                cityID[city1[0] - 'A'] = id;
                cityName[id++] = city1[0];
            }
            scanf("%s", city2);
            if(cityID[city2[0] - 'A'] == -1) {
                cityID[city2[0] - 'A'] = id;
                cityName[id++] = city2[0];
            }
            edges[i] = (Edge){cityID[city1[0] - 'A'], cityID[city2[0] - 'A']};
        }
        int **conn = (int**)malloc(id * sizeof(int*)), connSize[id];
        for(int i = 0; i < id; ++i) {
            conn[i] = NULL;
            connSize[i] = 0;
        }
        for(int i = 0; i < m; ++i) {
            conn[edges[i].parent] = (int*)realloc(conn[edges[i].parent], (connSize[edges[i].parent] + 1) * sizeof(int));
            conn[edges[i].parent][connSize[edges[i].parent]++] = edges[i].child;
            conn[edges[i].child] = (int*)realloc(conn[edges[i].child], (connSize[edges[i].child] + 1) * sizeof(int));
            conn[edges[i].child][connSize[edges[i].child]++] = edges[i].parent;
        }
        while(n--) {
            scanf("%s", city1);
            int start = cityID[city1[0] - 'A'];
            scanf("%s", city2);
            int end = cityID[city2[0] - 'A'];
            int prev[id];
            bool visited[id];
            memset(visited, 0, sizeof(visited));
            visited[curr[0] = prev[start] = start] = true;
            int currSize = 1, step = 0;
            while(!visited[end]) {
                int nextSize = 0;
                for(int i = 0; i < currSize; ++i) {
                    for(int j = 0; j < connSize[curr[i]]; ++j) {
                        if(!visited[conn[curr[i]][j]]) {
                            visited[next[nextSize++] = conn[curr[i]][j]] = true;
                            prev[conn[curr[i]][j]] = curr[i];
                        }
                    }
                }
                for(int i = 0; i < nextSize; ++i)
                    curr[i] = next[i];
                currSize = nextSize;
                ++step;
            }
            int route[step], i = 0;
            while(end != start) {
                route[i++] = end;
                end = prev[end];
            }
            putchar(cityName[start]);
            while(i--)
                putchar(cityName[route[i]]);
            putchar('\n');
        }
        if(cases)
            putchar('\n');
        for(int i = 0; i < id; ++i)
            free(conn[i]);
        free(conn);
    }
    return 0;
}
